Columns: 

- dual degree yes or no 'Q1'
- bullying yes or no 'Q31'
- job yes or no 'Q13'
- studying in your country yes or no 'Q4'
- discrimination or harrassment 'Q34'



### EDA

In [1]:
import pandas as pd

In [2]:
# load dataframe
df=pd.read_csv('data.csv',dtype='str')

In [3]:
df

ID.format  \
0     The published format which was employed   
1                            SNAP 2015 SHIFT    
2                            SNAP 2015 SHIFT    
3                            SNAP 2015 SHIFT    
4                            SNAP 2015 SHIFT    
...                                       ...   
6808                         SNAP 2015 SHIFT    
6809                         SNAP 2015 SHIFT    
6810                         SNAP 2015 SHIFT    
6811                         SNAP 2015 SHIFT    
6812                         SNAP 2015 SHIFT    

                            ID.completed  \
0     Case completed in Snap Interviewer   
1                              completed   
2                              completed   
3                              completed   
4                              completed   
...                                  ...   
6808                           completed   
6809                           completed   
6810                           completed   
6811                           completed   
6812                           completed   

                                            ID.language  \
0     What language would you like to complete the s...   
1                                               Spanish   
2                                               English   
3                                               English   
4                                               English   
...                                                 ...   
6808                                            English   
6809                                            English   
6810                                            English   
6811                                            English   
6812                                            English   

                     ID.site            ID.date                ID.start  \
0     Questionnaire location  Date of interview  Time interview started   
1                        NaN         06/14/2019                16:46:41   
2                        NaN         06/14/2019                17:10:02   
3                        NaN         06/14/2019                17:15:21   
4                        NaN         06/14/2019                16:51:28   
...                      ...                ...                     ...   
6808                     NaN         06/14/2019                16:51:46   
6809                     NaN         06/14/2019                16:54:50   
6810                     NaN         06/14/2019                16:58:30   
6811                     NaN         06/14/2019                15:10:17   
6812                     NaN         06/14/2019                17:11:57   

                        ID.endDate                ID.end  \
0     Completion date of interview  Time interview ended   
1                       06/14/2019              17:36:11   
2                       06/14/2019              17:38:25   
3                       06/14/2019              17:39:13   
4                       06/14/2019              17:46:58   
...                            ...                   ...   
6808                    06/14/2019              17:07:47   
6809                    06/14/2019              17:08:17   
6810                    06/14/2019              17:19:23   
6811                    06/14/2019              17:21:02   
6812                    06/14/2019              17:27:22   

                    ID.time  \
0     Duration of interview   
1                      49.5   
2                     28.38   
3                     23.87   
4                      55.5   
...                     ...   
6808                  16.02   
6809                  13.45   
6810                  20.88   
6811                 130.75   
6812                  15.42   

                                                     Q1  ...  \
0     Which, if any, of the following degrees are yo...  ...   
1                       Doctorate degree (PhD/DPhil/MD)  ...   
2                     

The first row is the questions.

In [4]:
# check the size of the dataset
df.shape

(6813, 274)

The dataset has 6812 data points and 274 columns. There are a lot of columns so our main goal is to clean it up.

In [5]:
# check column names
list(df.columns)

['ID.format',
 'ID.completed',
 'ID.language',
 'ID.site',
 'ID.date',
 'ID.start',
 'ID.endDate',
 'ID.end',
 'ID.time',
 'Q1',
 'Q1.a',
 'Q2',
 'Q3',
 'Q3.a',
 'Q4',
 'Q5',
 'Q6',
 'Q6.a',
 'Q7',
 'Q7.a',
 'Q8',
 'Q8.a',
 'Q9',
 'Q9.a',
 'Q10',
 'Q10.a',
 'Q11',
 'Q11.a',
 'Q12:1',
 'Q12:2',
 'Q12:3',
 'Q12:4',
 'Q12:5',
 'Q12:6',
 'Q12:7',
 'Q12:8',
 'Q12:9',
 'Q12:10',
 'Q12:11',
 'Q12.a',
 'Q13',
 'Q14',
 'Q14.a',
 'Q15.a',
 'Q15.b',
 'Q15.c',
 'Q15.d',
 'Q15.e',
 'Q15.f',
 'Q15.g',
 'Q15.h',
 'Q15.i',
 'Q15.j',
 'Q15.k',
 'Q15.l',
 'Q15.m',
 'Q15.n',
 'Q16',
 'Q17',
 'Q17.a',
 'Q18.a',
 'Q19.a',
 'Q20',
 'Q21.a',
 'Q21.b',
 'Q21.c',
 'Q21.d',
 'Q21.e',
 'Q21.f',
 'Q21.g',
 'Q21.h',
 'Q21.i',
 'Q22.a',
 'Q22.b',
 'Q22.c',
 'Q22.d',
 'Q22.e',
 'Q22.f',
 'Q22.g',
 'Q22.h',
 'Q22.i',
 'Q23',
 'Q24',
 'Q25',
 'Q25.a',
 'Q26',
 'Q27.a',
 'Q27.b',
 'Q27.c',
 'Q27.d',
 'Q27.e',
 'Q27.f',
 'Q28',
 'Q29',
 'Q29.a',
 'Q30.a',
 'Q30.b',
 'Q30.c',
 'Q30.d',
 'Q30.e',
 'Q30.f',
 'Q31',
 'Q32:1

Questions start from the 10th column.

In [6]:
# print out a list of the questions
question_list=list(df.columns)[9:]
for q in question_list:
    print('********')
    print(q)
    print(df[q][0])

********
Q1
Which, if any, of the following degrees are you currently studying for?
********
Q1.a
If other, please specify
********
Q2
Hidden
********
Q3
Which was the most important reason you decided to enrol in a PhD programme?
********
Q3.a
If other, please specify
********
Q4
Are you studying in the country you grew up in?   
********
Q5
Where do you currently live?
********
Q6
Which region in Asia?
********
Q6.a
If other, please specify
********
Q7
Which country in Australasia?
********
Q7.a
If other, please specify
********
Q8
Which country in Africa?
********
Q8.a
If other, please specify
********
Q9
Which country in Europe?
********
Q9.a
If other, please specify
********
Q10
Which country in North or Central America?
********
Q10.a
If other, please specify
********
Q11
Which country in South America?
********
Q11.a
If other, please specify
********
Q12:1
What prompted you to study outside your country of upbringing? 
********
Q12:2
What prompted you to study outside your count

#### Target

Our target column is Q28

In [7]:
# check values for target
df['Q28'].value_counts()

No                                                                          4083
Yes                                                                         2429
Prefer not to say                                                            292
Have you ever sought help for anxiety or depression caused by PhD study?       1
Name: Q28, dtype: int64

In [8]:
df['Q28'].value_counts(normalize=True)

No                                                                          0.600000
Yes                                                                         0.356943
Prefer not to say                                                           0.042910
Have you ever sought help for anxiety or depression caused by PhD study?    0.000147
Name: Q28, dtype: float64

There are 4083 Nos and 2429 Yeses. There's a little bit of class imbalance, but not too bad.

In [9]:
# check null values
df['Q28'].isna().sum()

8

There are not many "prefer not to say"s or null values, so we can safely drop them.

In [10]:
df=df.dropna(subset=['Q28'])
df

ID.format  \
0     The published format which was employed   
1                            SNAP 2015 SHIFT    
2                            SNAP 2015 SHIFT    
3                            SNAP 2015 SHIFT    
4                            SNAP 2015 SHIFT    
...                                       ...   
6808                         SNAP 2015 SHIFT    
6809                         SNAP 2015 SHIFT    
6810                         SNAP 2015 SHIFT    
6811                         SNAP 2015 SHIFT    
6812                         SNAP 2015 SHIFT    

                            ID.completed  \
0     Case completed in Snap Interviewer   
1                              completed   
2                              completed   
3                              completed   
4                              completed   
...                                  ...   
6808                           completed   
6809                           completed   
6810                           completed   
6811                           completed   
6812                           completed   

                                            ID.language  \
0     What language would you like to complete the s...   
1                                               Spanish   
2                                               English   
3                                               English   
4                                               English   
...                                                 ...   
6808                                            English   
6809                                            English   
6810                                            English   
6811                                            English   
6812                                            English   

                     ID.site            ID.date                ID.start  \
0     Questionnaire location  Date of interview  Time interview started   
1                        NaN         06/14/2019                16:46:41   
2                        NaN         06/14/2019                17:10:02   
3                        NaN         06/14/2019                17:15:21   
4                        NaN         06/14/2019                16:51:28   
...                      ...                ...                     ...   
6808                     NaN         06/14/2019                16:51:46   
6809                     NaN         06/14/2019                16:54:50   
6810                     NaN         06/14/2019                16:58:30   
6811                     NaN         06/14/2019                15:10:17   
6812                     NaN         06/14/2019                17:11:57   

                        ID.endDate                ID.end  \
0     Completion date of interview  Time interview ended   
1                       06/14/2019              17:36:11   
2                       06/14/2019              17:38:25   
3                       06/14/2019              17:39:13   
4                       06/14/2019              17:46:58   
...                            ...                   ...   
6808                    06/14/2019              17:07:47   
6809                    06/14/2019              17:08:17   
6810                    06/14/2019              17:19:23   
6811                    06/14/2019              17:21:02   
6812                    06/14/2019              17:27:22   

                    ID.time  \
0     Duration of interview   
1                      49.5   
2                     28.38   
3                     23.87   
4                      55.5   
...                     ...   
6808                  16.02   
6809                  13.45   
6810                  20.88   
6811                 130.75   
6812                  15.42   

                                                     Q1  ...  \
0     Which, if any, of the following degrees are yo...  ...   
1                       Doctorate degree (PhD/DPhil/MD)  ...   
2                     

In [11]:
# check if those rows are sucessfully dropped
df['Q28'].value_counts()

No                                                                          4083
Yes                                                                         2429
Prefer not to say                                                            292
Have you ever sought help for anxiety or depression caused by PhD study?       1
Name: Q28, dtype: int64

In [12]:
# drop "Prefer not to say"s
df=df[df['Q28']!='Prefer not to say']

In [13]:
# check if those rows are sucessfully dropped 
df['Q28'].value_counts()

No                                                                          4083
Yes                                                                         2429
Have you ever sought help for anxiety or depression caused by PhD study?       1
Name: Q28, dtype: int64

In [14]:
dict_target={'Yes': 1, 'No': 0}

In [15]:
df['Q28']=df['Q28'].map(dict_target)

<ipython-input-15-7a5e2d789d2b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Q28']=df['Q28'].map(dict_target)


In [16]:
df['Q28'].dtype

dtype('float64')

#### Features

In [17]:
def info(label):
    '''
    input string of column name
    output comprehensive analysis of the column
    '''
    print(f'column name: {label}')
    print('     ')
    print(f'Question: {df[label][0]}')
    print('     ')
    print(f'number of null values: {df[label].isna().sum()}')
    print('     ')
    print('Value Count:')
    
    print(df[label].value_counts())

In [18]:
# check Q1
info('Q1')

column name: Q1
     
Question: Which, if any, of the following degrees are you currently studying for?
     
number of null values: 0
     
Value Count:
Doctorate degree (PhD/DPhil/MD)                                            6294
Dual doctorate degree (MD-PhD, PhD-PhD or other combination)                218
Which, if any, of the following degrees are you currently studying for?       1
Name: Q1, dtype: int64


In [19]:
dict_Q1={'Doctorate degree (PhD/DPhil/MD)': 0, 
         'Dual doctorate degree (MD-PhD, PhD-PhD or other combination)': 1, 
         'Which, if any, of the following degrees are you currently studying for?':'dual_degree'}

In [20]:
df['Q1']=df['Q1'].map(dict_Q1)

<ipython-input-20-53e31f3f75ed>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Q1']=df['Q1'].map(dict_Q1)


In [21]:
df['Q1'].value_counts()

0              6294
1               218
dual_degree       1
Name: Q1, dtype: int64

In [22]:
info('Q4')

column name: Q4
     
Question: Are you studying in the country you grew up in?   
     
number of null values: 0
     
Value Count:
Yes                                                   4153
 No                                                   2359
Are you studying in the country you grew up in?          1
Name: Q4, dtype: int64


In [23]:
dict_Q4={'Yes':1, ' No':0, 'Are you studying in the country you grew up in? \xa0 ': 'home country'}

In [24]:
df['Q4']=df['Q4'].map(dict_Q4)

<ipython-input-24-80b37a29571b>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Q4']=df['Q4'].map(dict_Q4)


In [25]:
df['Q4'].value_counts()

1               4153
0               2359
home country       1
Name: Q4, dtype: int64

In [26]:
info('Q13')

column name: Q13
     
Question: Do you have a job alongside your studies?
     
number of null values: 0
     
Value Count:
No                                           5249
Yes                                          1263
Do you have a job alongside your studies?       1
Name: Q13, dtype: int64


In [27]:
df['Q13'].value_counts()

No                                           5249
Yes                                          1263
Do you have a job alongside your studies?       1
Name: Q13, dtype: int64

In [28]:
df["Q13"][21]

'Yes '

In [29]:
dict_Q13={'Yes ':1, 'No':0, 'Do you have a job alongside your studies?': 'job'}

In [30]:
df['Q13']=df['Q13'].map(dict_Q13)

<ipython-input-30-aeb4f6ca4191>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Q13']=df['Q13'].map(dict_Q13)


In [31]:
info('Q31')

column name: Q31
     
Question: Do you feel that you have experienced bullying in your PhD program?
     
number of null values: 7
     
Value Count:
No                                                                     4902
Yes                                                                    1388
Prefer not to say                                                       215
Do you feel that you have experienced bullying in your PhD program?       1
Name: Q31, dtype: int64


In [32]:
dict_Q31={'Yes':1, 'No': 0, 'Prefer not to say': 2, 'Do you feel that you have experienced bullying in your PhD program?': 'bullying'}

In [33]:
df['Q31']=df['Q31'].map(dict_Q31)

<ipython-input-33-9b6da20be755>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Q31']=df['Q31'].map(dict_Q31)


In [34]:
info('Q34')

column name: Q34
     
Question: Do you feel that you have experienced discrimination or harassment in your PhD program?
     
number of null values: 24
     
Value Count:
No                                                                                         4937
Yes                                                                                        1334
Prefer not to say                                                                           217
Do you feel that you have experienced discrimination or harassment in your PhD program?       1
Name: Q34, dtype: int64


In [35]:
dict_Q34={'Yes':1, 'No': 0, 'Prefer not to say': 2, 'Do you feel that you have experienced discrimination or harassment in your PhD program?': 'discrimination/harrassment'}

In [36]:
df['Q34']=df['Q34'].map(dict_Q34)

<ipython-input-36-d297d21c6dae>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Q34']=df['Q34'].map(dict_Q34)


In [37]:
info('Q19.a')

column name: Q19.a
     
Question: How satisfied are you with your PhD experience?
     
number of null values: 0
     
Value Count:
6                                                  2085
5                                                  1908
4 = Neither satisfied nor dissatisfied              717
3                                                   659
7 = Extremely satisfied                             592
2                                                   360
1 = Not at all satisfied                            191
How satisfied are you with your PhD experience?       1
Name: Q19.a, dtype: int64


Satisfaction is on a scale of 1 to 7, not at all satisfied to extremely satisfied. So the values just need to be reduced to numbers. 

In [38]:
df['Q19.a'][1:]=df['Q19.a'][1:].map(lambda x:int(x[0]))

<ipython-input-38-a6aa1b63e538>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Q19.a'][1:]=df['Q19.a'][1:].map(lambda x:int(x[0]))
/Users/elylin/opt/anaconda3/envs/learn-env/lib/python3.8/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)
/Users/elylin/opt/anaconda3/envs/learn-env/lib/python3.8/site-packages/pandas/core/series.py:1043: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#r

In [39]:
df['Q19.a'][0]='satisfaction'

<ipython-input-39-946e16614388>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Q19.a'][0]='satisfaction'


In [40]:
df['Q19.a'].value_counts()

6               2085
5               1908
4                717
3                659
7                592
2                360
1                191
satisfaction       1
Name: Q19.a, dtype: int64

In [41]:
info('Q24')

column name: Q24
     
Question: On average, how many hours a week do you typically spend on your PhD programme?
     
number of null values: 0
     
Value Count:
41-50 hours                                                                        1785
51-60 hours                                                                        1609
31-40 hours                                                                         783
61-70 hours                                                                         774
71-80 hours                                                                         425
More than 80 hours                                                                  328
Less than 11 hours                                                                  287
21-30 hours                                                                         274
11-20 hours                                                                         247
On average, how many hours a week do you typi

For this column, we can just take the middle of the range and cap it at 81 hours for simplicity. 

In [42]:
dict_Q24={'41-50 hours': 45, 
          '51-60 hours': 55, 
          '31-40 hours': 35, 
          '61-70 hours': 65, 
          '71-80 hours': 75, 
          'More than 80 hours':81, 
          'Less than 11 hours': 5, 
          '21-30 hours': 25, 
          '11-20 hours': 15, 
          'On average, how many hours a week do you typically spend on your PhD programme?': 'hours_worked'}

In [43]:
df['Q24']=df['Q24'].map(dict_Q24)

<ipython-input-43-894a6e5e1440>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Q24']=df['Q24'].map(dict_Q24)


In [44]:
df['Q24'].value_counts()

45              1785
55              1609
35               783
65               774
75               425
81               328
5                287
25               274
15               247
hours_worked       1
Name: Q24, dtype: int64

In [45]:
info('Q57')

column name: Q57
     
Question: Are you…
     
number of null values: 0
     
Value Count:
Female (including trans female)    3298
Male (including trans male)        3149
Prefer not to say                    35
Gender queer / Non binary            30
Are you…                              1
Name: Q57, dtype: int64


In [46]:
dict_Q57={'Female (including trans female)':0, 'Male (including trans male)': 1, 'Prefer not to say': 2, 'Gender queer / Non binary': 3, 'Are you…': 'gender'}

In [47]:
df['Q57']=df['Q57'].map(dict_Q57)

<ipython-input-47-3a54b3feb3ac>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Q57']=df['Q57'].map(dict_Q57)


In [48]:
df['Q57'].value_counts()

0         3298
1         3149
2           35
3           30
gender       1
Name: Q57, dtype: int64

In [79]:
info('Q56')

column name: Q56
     
Question: What is your age?
     
number of null values: 0
     
Value Count:
25 - 34              5019
18 - 24               789
35 - 44               530
45 - 54               103
55 - 64                35
Prefer not to say      23
65 or over             13
What is your age?       1
Name: Q56, dtype: int64


In [49]:
simple_df=df[['Q1', 'Q4', 'Q13', 'Q31', 'Q34', 'Q19.a', 'Q24', 'Q57']]

In [50]:
# make the first row column names
simple_df.columns = simple_df.iloc[0] 

# drop the first row
simple_df = simple_df.drop([0])


In [51]:
simple_df

0    dual_degree home country job bullying discrimination/harrassment  \
1              0            1   0        0                          1   
2              0            0   0        0                          1   
3              0            1   0        0                          0   
5              0            1   0        0                          0   
6              0            1   0        0                          0   
...          ...          ...  ..      ...                        ...   
6806           0            1   0        0                          0   
6807           0            0   0        0                          1   
6809           0            1   0        0                          0   
6811           0            1   0        0                          0   
6812           0            0   0        0                          0   

0    satisfaction hours_worked gender  
1               5           55      0  
2               6           65      1  
3               6           65      1  
5               3           55      1  
6               4           75      0  
...           ...          ...    ...  
6806            6           45      1  
6807            6           75      1  
6809            4           35      0  
6811            6           75      1  
6812            5           45      0  

[6512 rows x 8 columns]

In [52]:
simple_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6512 entries, 1 to 6812
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   dual_degree                 6512 non-null   object
 1   home country                6512 non-null   object
 2   job                         6512 non-null   object
 3   bullying                    6505 non-null   object
 4   discrimination/harrassment  6488 non-null   object
 5   satisfaction                6512 non-null   object
 6   hours_worked                6512 non-null   object
 7   gender                      6512 non-null   object
dtypes: object(8)
memory usage: 457.9+ KB


In [53]:
# fill na for 'job', 'bullying', and 'discrimination/harrassment'
# 
# assume all nulls are 'prefer not to say'
simple_df['bullying'].fillna(value=2,inplace=True)
simple_df['discrimination/harrassment'].fillna(value=2,inplace=True)

In [54]:
simple_df=simple_df.astype(int)

### Train Test Split

In [55]:
from sklearn.model_selection import train_test_split

In [56]:
X=simple_df
y=df['Q28'][1:]

In [57]:
X_tr, X_test, y_tr, y_test = train_test_split(X, y, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_tr, y_tr, random_state=42)

### Modeling

#### Baseline Understanding: 
predict majority class

In [58]:
y_train.value_counts(normalize=True)

0.0    0.62326
1.0    0.37674
Name: Q28, dtype: float64

If we predict the majority class of not going to need help for anxiety/depression as a result of PhD study (assuming that if people need help, they ask for help), we will be right 62.3% of the time.

#### Logistic Regression

In [59]:
from sklearn.linear_model import LogisticRegression

In [60]:
lr=LogisticRegression()
lr.fit(X_train,y_train)

LogisticRegression()

In [61]:
print(lr.score(X_train,y_train))
print(lr.score(X_val,y_val))

0.6691236691236692
0.6568386568386568


#### Decision Tree

In [62]:
from sklearn.tree import DecisionTreeClassifier

In [63]:
dt=DecisionTreeClassifier(max_depth=7,random_state=42)

In [64]:
dt.fit(X_train,y_train)

DecisionTreeClassifier(max_depth=7, random_state=42)

In [65]:
print(dt.score(X_train,y_train))
print(dt.score(X_val,y_val))

0.7037947037947038
0.6478296478296478


So for decision tree there is only 7(?) decisions to make, and it cannot possibly improve further, add a graph to show that later.

#### Naive Bayes

In [66]:
from sklearn.naive_bayes import CategoricalNB

In [67]:
nb=CategoricalNB()

In [68]:
nb.fit(X_train,y_train)

CategoricalNB()

In [69]:
print(nb.score(X_train,y_train))
print(nb.score(X_val,y_val))

0.6773136773136773
0.6609336609336609


#### CatBoost

In [70]:
from catboost import CatBoostClassifier

In [71]:
cb=CatBoostClassifier(silent=True)

In [72]:
cb.fit(X_train,y_train)

In [73]:
print(cb.score(X_train,y_train))
print(cb.score(X_val,y_val))

0.7474747474747475
0.6461916461916462


#### XGBoost

In [74]:
from xgboost import XGBClassifier

In [75]:
from sklearn.model_selection import GridSearchCV

In [76]:
xgb=XGBClassifier()

In [77]:
xgb.fit(X_train,y_train)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.300000012, max_delta_step=0, max_depth=6,
              min_child_weight=1, missing=nan, monotone_constraints='()',
              n_estimators=100, n_jobs=0, num_parallel_tree=1, random_state=0,
              reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [78]:
print(xgb.score(X_train,y_train))
print(xgb.score(X_val,y_val))

0.7633087633087633
0.6461916461916462


Basically we didn't give the model enough information to go off on. We need to add more information to the models.